File: core/message_parser.py
import re
from rapidfuzz import fuzz

PROJECTS = ["Top Terrace", "9TH FLOOR (CENTRIC)"]

TASK_KEYWORDS = {
    "Top Terrace": ["tile", "membrane", "malba", "slope", "solar", "waterproof", "ponding"],
    "9TH FLOOR (CENTRIC)": ["tile", "malba", "slope", "waterproof", "ponding"]
}

def detect_project(text: str) -> str:
    best_match = None
    highest_score = 0
    
    for project in PROJECTS:
        score = fuzz.partial_ratio(project.lower(), text.lower())
        if score > highest_score:
            highest_score = score
            best_match = project
            
    if highest_score >= 75:
        return best_match
    return None

def detect_task(text: str, project_name: str) -> str:
    if not project_name or project_name not in TASK_KEYWORDS:
        return None
        
    keywords = TASK_KEYWORDS[project_name]
    best_match = None
    highest_score = 0
    words = text.lower().split()
    
    for keyword in keywords:
        for word in words:
            score = fuzz.partial_ratio(keyword, word)
            if score > highest_score:
                highest_score = score
                best_match = keyword
                
    if highest_score >= 70:
        return best_match
    return None

def extract_progress(text: str):
    match = re.search(r'(\d{1,3})\s*%', text)
    if match:
        val = int(match.group(1))
        if 0 <= val <= 100:
            return val
    return None

def extract_blocker(text: str, parsed_fields: dict) -> str:
    clean_text = text
    
    # Remove project
    if parsed_fields.get("project"):
        # naive replace ignoring case
        pattern = re.compile(re.escape(parsed_fields["project"]), re.IGNORECASE)
        clean_text = pattern.sub('', clean_text)
        
    # Remove task keyword
    if parsed_fields.get("task_keyword"):
        pattern = re.compile(re.escape(parsed_fields["task_keyword"]), re.IGNORECASE)
        clean_text = pattern.sub('', clean_text)
        
    # Remove progress
    progress_val = parsed_fields.get("progress")
    if progress_val is not None:
        clean_text = re.sub(rf'{progress_val}\s*%', '', clean_text)
        
    # Extra cleanups like "done", spaces, commas
    clean_text = re.sub(r'\b(done|percent|completed)\b', '', clean_text, flags=re.IGNORECASE)
    clean_text = clean_text.replace(',', '').strip()
    
    if not clean_text:
        return None
    return ' '.join(clean_text.split())

def calculate_confidence(parsed_dict: dict) -> str:
    has_proj = bool(parsed_dict.get("project"))
    has_task = bool(parsed_dict.get("task_keyword"))
    has_prog = parsed_dict.get("progress") is not None
    
    if has_proj and has_task and has_prog:
        return "high"
    elif has_task and has_prog:
        return "medium"
    return "low"

def parse_message(text: str) -> dict:
    project = detect_project(text)
    task_keyword = detect_task(text, project)
    progress = extract_progress(text)
    
    parsed = {
        "project": project,
        "task_keyword": task_keyword,
        "progress": progress
    }
    
    blocker = extract_blocker(text, parsed)
    parsed["blocker"] = blocker
    
    confidence = calculate_confidence(parsed)
    parsed["confidence"] = confidence
    
    return parsed

File: simulator/lib/messageParser.js
import Fuse from 'fuse.js';

const PROJECTS = [
    'Top Terrace',
    '9TH FLOOR (CENTRIC)'
];

const TASKS = [
    'tile', 'membrane', 'malba', 'slope', 'solar', 'waterproof', 'ponding'
];

export function parseUpdateMessage(message) {
    if (!message) return { confidence: 'Low', missing: ['message'] };

    const parsed = {
        project: null,
        task: null,
        progress: null,
        blocker: null,
        confidence: 'Low'
    };

    // Extract progress % using regex
    const progressMatch = message.match(/(\d+)%\s*(?:completed|done)?/i) || message.match(/(\d+)\s*percent/i);
    if (progressMatch) {
        parsed.progress = parseInt(progressMatch[1], 10);
    }

    // Extract project using Fuse.js
    const projectFuse = new Fuse(PROJECTS, { includeScore: true, threshold: 0.4 });
    const projectHits = projectFuse.search(message);
    if (projectHits.length > 0) {
        parsed.project = projectHits[0].item;
    }

    // Extract task using Fuse.js
    const taskFuse = new Fuse(TASKS, { includeScore: true, threshold: 0.4 });
    // Tokenize message to check each word for tasks to avoid missing them in a long string
    const words = message.split(/[\s,.-]+/);
    let bestTask = null;
    let bestTaskScore = 1;
    for (const word of words) {
        const hits = taskFuse.search(word);
        if (hits.length > 0 && hits[0].score < bestTaskScore) {
            bestTaskScore = hits[0].score;
            bestTask = hits[0].item;
        }
    }
    if (!bestTask) {
        // try matching the whole message if words didn't catch it
        const tkHits = taskFuse.search(message);
        if (tkHits.length > 0) {
            bestTask = tkHits[0].item;
        }
    }
    parsed.task = bestTask;

    // Extract blocker (simple heuristic: look for sentences or keywords after "delay", "blocked", "issue")
    const blockerMatch = message.match(/(?:delay|blocked|issue|blocker)[:\s]+([^.]+)/i);
    if (blockerMatch) {
        parsed.blocker = blockerMatch[1].trim();
    }

    // Determine confidence
    if (parsed.project && parsed.task && parsed.progress !== null) {
        parsed.confidence = 'High';
    } else if (parsed.task && parsed.progress !== null) {
        parsed.confidence = 'Medium';
    } else {
        parsed.confidence = 'Low';
    }

    return parsed;
}
